### ISSUE TITLE
Event Creator blows up on python3

### ISSUE THREAD

**Issue by Half-Shot** [matrix-org/synapse#3866]
Running develop (with py3) on the ``event_creator`` worker produces this little message and causes the thing to exit:

```
<Half-Shot> Sep 13 20:08:14 Halfy-VM python[4085]:   File "/srv/matrix/synapse/lib/python3.6/site-packages/synapse/storage/user_directory.py", line 577, in get_current_state_deltas
<Half-Shot> Sep 13 20:08:14 Halfy-VM python[4085]:     if not self._curr_state_delta_stream_cache.has_any_entity_changed(prev_stream_id):
<Half-Shot> Sep 13 20:08:14 Halfy-VM python[4085]: AttributeError: 'EventCreatorSlavedStore' object has no attribute '_curr_state_delta_stream_cache'
```

There is more to the logs, but that seemed like the critical bit to grab at the time. I can dig further if needed.

**Comment by Half-Shot** [comment]
A bettter stack trace:
```
2018-09-17 15:30:23,760 - synapse.replication.tcp.protocol - 278 - DEBUG - - [anon-oaQfk] Queing as conn 'connecting', cmd: <synapse.replication.tcp.commands.NameCommand object at 0x7f275af4ae10>
2018-09-17 15:30:23,760 - synapse.replication.tcp.protocol - 153 - INFO - - [anon-oaQfk] Connection established
2018-09-17 15:30:23,760 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'caches' from 15474444
2018-09-17 15:30:23,760 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'public_rooms' from 337
2018-09-17 15:30:23,760 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'events' from 10758359
2018-09-17 15:30:23,761 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'backfill' from 96600
2018-09-17 15:30:23,761 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'device_lists' from 326175
2018-09-17 15:30:23,761 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'push_rules' from 648
2018-09-17 15:30:23,761 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'receipts' from 7774672
2018-09-17 15:30:23,761 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'pushers' from 30098
2018-09-17 15:30:23,761 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'tag_account_data' from 35663
2018-09-17 15:30:23,761 - synapse.replication.tcp.protocol - 578 - INFO - - [anon-oaQfk] Subscribing to replication stream: 'account_data' from 35663
2018-09-17 15:30:23,768 - synapse.metrics - 268 - INFO - - Collecting gc 0
2018-09-17 15:30:24,703 - synapse.metrics - 268 - INFO - - Collecting gc 1
2018-09-17 15:30:24,718 - twisted - 243 - CRITICAL - - Unhandled error in Deferred:
2018-09-17 15:30:24,718 - twisted - 243 - CRITICAL - - 
Traceback (most recent call last):
  File "/srv/matrix/synapse/lib/python3.6/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/srv/matrix/synapse/lib/python3.6/site-packages/synapse/storage/_base.py", line 377, in runWithConnection
    defer.returnValue(result)
  File "/srv/matrix/synapse/lib/python3.6/site-packages/twisted/internet/defer.py", line 1362, in returnValue
    raise _DefGen_Return(val)
twisted.internet.defer._DefGen_Return: 10539556

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/srv/matrix/synapse/lib/python3.6/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/srv/matrix/synapse/lib/python3.6/site-packages/synapse/storage/_base.py", line 335, in runInteraction
    defer.returnValue(result)
  File "/srv/matrix/synapse/lib/python3.6/site-packages/twisted/internet/defer.py", line 1362, in returnValue
    raise _DefGen_Return(val)
twisted.internet.defer._DefGen_Return: 10539556

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/srv/matrix/synapse/lib/python3.6/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/srv/matrix/synapse/lib/python3.6/site-packages/twisted/python/failure.py", line 491, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/srv/matrix/synapse/lib/python3.6/site-packages/synapse/handlers/user_directory.py", line 113, in notify_new_event
    yield self._unsafe_process()
  File "/srv/matrix/synapse/lib/python3.6/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/srv/matrix/synapse/lib/python3.6/site-packages/synapse/handlers/user_directory.py", line 151, in _unsafe_process
    deltas = yield self.store.get_current_state_deltas(self.pos)
  File "/srv/matrix/synapse/lib/python3.6/site-packages/synapse/storage/user_directory.py", line 577, in get_current_state_deltas
    if not self._curr_state_delta_stream_cache.has_any_entity_changed(prev_stream_id):
AttributeError: 'EventCreatorSlavedStore' object has no attribute '_curr_state_delta_stream_cache'
```

**Comment by hawkowl** [comment]
Fixed in #4027
